bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs

```diff
@@ -340,35 +344,52 @@ fn userdata(
     endpoint: &Option<String>,
     certificate: &Option<String>,
     cluster_dns_ip: &Option<String>,
+    custom_user_data: &Option<CustomUserData>,
     memo: &ProductionMemo,
 ) -> ProviderResult<String> {
-    Ok(match cluster_type {
-        ClusterType::Eks => base64::encode(format!(
-            r#"[settings.updates]
-ignore-waves = true
-    
-[settings.kubernetes]
-api-server = "{}"
-cluster-name = "{}"
-cluster-certificate = "{}"
-cluster-dns-ip = "{}""#,
-            endpoint
-                .as_ref()
-                .context(memo, "Server endpoint is required for eks clusters.")?,
-            cluster_name,
-            certificate
-                .as_ref()
-                .context(memo, "Cluster certificate is required for eks clusters.")?,
-            cluster_dns_ip
-                .as_ref()
-                .context(memo, "Cluster DNS IP is required for eks clusters.")?,
-        )),
-        ClusterType::Ecs => base64::encode(format!(
-            r#"[settings.ecs]
-cluster = "{}""#,
-            cluster_name,
-        )),
-    })
+    let userdata = match custom_user_data {
```

Comment:
I would simplify this match case by early returning if there is no custom userdata. Something like 

```rust
let default_userdata = match cluster_type {
    ClusterType::Eks => default_eks_userdata(x, y, z),
    ClusterType::Ecs => default_ecs_userdata(x),
}

let custom_userdata = if let Some(value) = customer_user_data {
    value
} else {
    return Ok(default_userdata);
};

match custom_user_data {
    /* ... */
}
```

Note: in the case of `Replace` the `default_userdata` will have been constructed unnecessarily, but the easy-to-read program flow outweighs the construction of an unused string here.